File: ruby/abc218_d.rb
# https://atcoder.jp/contests/abc218/tasks/abc218_d
# https://atcoder.jp/contests/abc218/submissions/25780847 # スマート

def to_key(point)
  "#{point[0]},#{point[1]}"
end

n = gets.chomp.to_i

memo = {}
points = n.times.map {
  point = gets.chomp.split(" ").map(&:to_i)
  memo[to_key(point)] = true
  point
}

count = 0
points.combination(2).each { |p1, p2|
  next if p1[0] == p2[0] || p1[1] == p2[1]
  p3 = [p1[0], p2[1]]
  p4 = [p2[0], p1[1]]

  count += 1 if memo[to_key(p3)] && memo[to_key(p4)]
}
puts count / 2

File: rust/src/abc218_d.rs
// https://atcoder.jp/contests/abc218/tasks/abc218_d

use proconio::input;
use std::collections::HashMap;

fn main() {
    input! {
        n: usize,
        points: [(usize, usize); n]
    }
    let mut map = HashMap::new();
    for point in points.iter() {
        map.insert(point, true);
    }
    let mut ans: usize = 0;
    for i in 0..n {
        for j in i + 1..n {
            if !(points[i].0 == points[j].0 || points[i].1 == points[j].1) {
                let p3 = (points[i].0, points[j].1);
                let p4 = (points[j].0, points[i].1);
                if map.contains_key(&p3) && map.contains_key(&p4) {
                    ans += 1;
                }
            }
        }
    }
    println!("{}", ans / 2);
}
